### PR TITLE
[Alpha WIP] Web Support

### PR DESCRIPTION
**Pull request by stevensJourney**
# Description 

NOTE: This is currently an alpha feature. This should be released as a prerelease once merged into the `alpha-web-prerelease` branch.

Currently the `SqliteDatabase` and other implementations heavily rely on Dart isolates and the Dart FFI Sqlite implementation. 

Importing this package into a Flutter web application will cause compilation to fail due to these dependencies not being available on Web. 

This PR adds abstraction to the database connection and open factory classes. Implementations are split between native FFI platforms and Web. Library exports are conditionally managed in order to maintain compatibility with the current library API and provide web implementations for web projects.

The web implementation uses the [Drift]https://pub.dev/packages/drift package to provide asynchronous WASM sqlite operations. Reasons for using Drift:

- Provides functionality built upon the standard Sqlite3.dart package
- Provides automatic shared/dedicated Webworker support fallbacks with automatic VFS support detection and selection
- Implements async message protocols through webworker proxies. 

# Notes
Web database connections run via a single connection to a web worker. This prevents direct access to the `CommonDatabase` instance. 

Due to this, the following differences are in web:
- Web's `OpenFactory.open()` Has the ability to use a synchronous `CommonDatabase` connection which is instantiated with standard `sqlite3.dart` and an `IndexDB` VFS. This VFS does not use the same implementation that Drift uses.  It is not possible to use these direct connections in tandem with Drift async connections. The functionality is just provided for API compatibility, but might be removed in future - or potentially updated to use the Drift VFS.
- `.computeWithDatabase(...` is not availble on Web. use thereof will throw an `UnimplementedError`.
-  Web uses a single database connection for reads and write. Due to this web currently does not error if read or write operations are done within the opposite lock or transaction context.

# Status

## Drift
Watched queries and table change updates rely on Drift worker changes which are implemented here https://github.com/powersync-ja/drift/pull/1 . This Fork is not currently a published package. In order to publish alpha versions of this package, the standard Drift dependency is used. Workers which provide watched queries need to be compiled using the linked Drift repo.

Tests for watched queries have been disabled on Web for now. They have been verified manually.

## Multiple Tabs
The Drift connection supports shared webworkers which is good for multiple tab support. But this current implementation uses a non-shared in-memory mutex lock for read and write locks. This means locks are currently not global between tabs. Future work will use Web `navigator.locks` bindings to provide mutex locking. 

## Package publishing
The `Build` test currently fails due to a low publish score caused by the version of `package:js^0.6.3` The Sqlite3 library requires this version, bumping to the latest stable causes package resolving issues.
image

## Related PRs
This PR is required for https://github.com/powersync-ja/powersync.dart/pull/56

### CODE REVIEW

File: lib/definitions.dart [hunk 1]
```diff
@@ -0,0 +1,8 @@
+export 'package:sqlite_async/src/update_notification.dart';
```
Comment (rkistner) [inline lib/definitions.dart:1]: What's the purpose of this file / these exports? A large portion of these are already part of the sqlite_async.dart exports

File: lib/src/web/database/web_db_context.dart [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    return db.select(sql, parameters);
+  }
+
+  @override
+  Future<Row?> getOptional(String sql,
+      [List<Object?> parameters = const []]) async {
+    try {
+      return (await getAll(sql, parameters)).first;
+    } catch (ex) {
+      return null;
```
Comment (rkistner) [inline lib/src/web/database/web_db_context.dart:46]: There could be exceptions other than "no results" caught here. Rather use an implementation like this:

https://github.com/powersync-ja/sqlite_async.dart/blob/fdd09c34475dbd5aa6681c4d3af1dadb5c9732b8/lib/src/native/database/native_sqlite_connection_impl.dart#L226

File: lib/src/common/connection/sync_sqlite_connection.dart [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    return db.select(sql, parameters);
+  }
+
+  @override
+  Future<Row?> getOptional(String sql,
+      [List<Object?> parameters = const []]) async {
+    try {
+      return db.select(sql, parameters).first;
+    } catch (ex) {
+      return null;
```
Comment (rkistner) [inline lib/src/common/connection/sync_sqlite_connection.dart:83]: Same here.

**Review by rkistner — COMMENTED** [review]
Is it feasible to use a similar naming structure to https://github.com/powersync-ja/powersync.dart/pull/56, where the base classes are just called "SqliteDatabase", and the implementations "SqliteDatabaseImpl"? (Same for Mutex)

**Comment by stevensJourney** [comment]
> Is it feasible to use a similar naming structure to [powersync-ja/powersync.dart#56](https://github.com/powersync-ja/powersync.dart/pull/56), where the base classes are just called "SqliteDatabase", and the implementations "SqliteDatabaseImpl"? (Same for Mutex)

To some extent it is possible. The linked naming structure has been applied to `SqliteDatabase` and `IsolateConnectionFactory`. It was not a straight forward task to apply to `DefaultOpenFactory` due to it's nature of being extended externally.

File: lib/definitions.dart (hunk 1 above)
Comment (stevensJourney) [reply lib/definitions.dart]: This was for testing in the early stages of splitting the different platform implementations. It's no longer needed and has been deleted. Thanks for spotting.

File: lib/src/common/connection/sync_sqlite_connection.dart (hunk 3 above)
Comment (stevensJourney) [reply lib/src/common/connection/sync_sqlite_connection.dart]: Implemented as above.

**Review by rkistner — APPROVED** [review]
You can temporarily bump the `--exit-code-threshold` to 10 to let the tests pass.

**Review by rkistner — APPROVED** [review]

**Pull request merged**